packages/lib-user/src/components/GroupStats/components/MembersList/MembersListContainer.js

```diff
@@ -0,0 +1,78 @@
+import { shape, string } from 'prop-types'
+import { useEffect, useState } from 'react'
+
+import { usePanoptesMemberships } from '@hooks'
+
+import {
+  deletePanoptesMembership,
+  updatePanoptesMembership
+} from '@utils'
+
+import MembersList from './MembersList'
+
+function MembersListContainer({
+  authUser,
+  group
+}) {
+  const [memberships, setMemberships] = useState([])
+
+  const {
+    data: membershipsData,
+    error,
+    isLoading
+  } = usePanoptesMemberships({
+    authUserId: authUser?.id,
+    query: {
+      include: 'user',
+      user_group_id: group?.id
+    }
+  })
+
+  useEffect(() => {
+    if (membershipsData) {
+      setMemberships(membershipsData.memberships.filter(membership => membership.state === 'active'));
+    }
+  }, [membershipsData]);
+
+  async function handleDeleteMembership({ membershipId }) {
+    const deleteMembershipResponse = await deletePanoptesMembership({ membershipId })
+    if (deleteMembershipResponse.ok) {
+      const updatedMemberships = memberships.filter(membership => membership.id !== membershipId)
+      setMemberships(updatedMemberships)
+    }
+  }
+
+  async function handleUpdateMembership({ membershipId, data }) {
+    const updateMembershipResponse = await updatePanoptesMembership({ membershipId, data })
+    if (updateMembershipResponse.ok) {
+      const updatedMemberships = memberships.map(membership => {
+        if (membership.id === membershipId) {
+          membership.roles = data.roles
+        }
+        return membership
+      })
+      setMemberships(updatedMemberships)
```

Comment:
I'm not exactly sure how I'd implement it using `membershipsData` as state, as I think I'd have to add a param to the `usePanoptesMemberships` hook so it knew to update? Open to suggestion, but this was what felt easiest initially. Definitely have to refactor for sad paths in a few places, maybe worth creating a dedicated card/issue.

Reply:
I thought about this concept a bit more, and I think it's worth looking into [SWR's mutations](https://swr.vercel.app/docs/mutation). This [youtube tutorial](https://www.youtube.com/watch?v=qURCyd3ycfw&ab_channel=CodeGenix) is also a great example of SWR data fetching + mutation in practice.

The majority of data handling in the user groups and stats pages simply fetches data from panoptes, which you've implemented with SWR hooks. This works well because the fetched/cached data can't be updated on the user groups or stats pages (for instance, stats are modified on classify pages or a user's display name is updated on their profile settings page). However, group memberships is the major feature that requires modifying data in panoptes while also updating the currently-viewed UI on the fly.

SWR provides hook `useSWRMutation` for cases where you need to POST, PUT, or DEL to the backend, yet continue to use the data returned from a GET hook like `usePanoptesMemberships` in MembersListContainer. `useSWRMutation` is an optimistic UI pattern, but without creating a separate `useState` variable. 

The second reason I think it's worth looking into SWR mutations is to handle the forced page refresh I commented on. Forcing a page refresh after modifying group membership data feels like an anti-pattern in a React application where components in the UI should re-render when passed new data props. GroupStatsContainer also utilizes `usePanoptesMemberships` and passes `membershipsData` to GroupStats. With SWR mutations you could rely on `membershipsData` to update after a mutation and cause a GroupStats re-render without a total page refresh.

Please let me know your thoughts! And if you'd like to separate changes to membership data handling into another PR or continue on this one.

Reply:
The `useSWRMutation` SWR hook noted sounds great to me, certainly preferable. My only hesitation is adding to this PR's changes, but if that's ok with you then it's ok with me and I'd like to try refactoring accordingly.

Other user/group stats PRs aren't dependent on this PR so it won't hold anything up if this gets additional work.

Reply:
Feel free to add to this PR if easiest in the flow of features for lib-user! You could also create a PR requesting to merge into this one with isolated changes for data mutation. Happy to review either way.